src/services/logger/logger.service.js

```diff
@@ -84,17 +84,103 @@ angular.module('vlui')
       SPEC_PREVIEW_DISABLED: {category:'PREVIEW', id: 'SPEC_PREVIEW_DISABLED', level: service.levels.INFO}
     };
 
+    // create noop service if websql is not supported
+    if ($window.openDatabase === undefined) {
+      console.warn('No websql support and thus no logging.');
+      service.logInteraction = function() {};
+      return service;
+    }
+
+    // get user id once in the beginning
+    var user = $location.search().user;
+
+    service.db = $webSql.openDatabase('logs', '1.0', 'Logs', 2 * 1024 * 1024);
+
+    service.tableName = 'Logs_' + consts.appId;
+
+    // (zening) TODO: check if the table is correct, do we really need time? will time be automatically added?
+    service.createTableIfNotExists = function() {
+      service.db.createTable(service.tableName, {
+        'userid': {
+          'type': 'INTEGER',
+          'null': 'NOT NULL'
+        },
+        'time': {
+          'type': 'TIMESTAMP',
+          'null': 'NOT NULL'
+        },
+        'actionCategory': {
+          'type': 'TEXT',
+          'null': 'NOT NULL'
+        },
+        'actionId': {
+          'type': 'TEXT',
+          'null': 'NOT NULL'
+        },
+        'label': {
+          'type': 'TEXT',
+          'null': 'NOT NULL'
+        },
+        'data': {
+          'type': 'TEXT'
+        }
+      });
+    };
+
+    service.clear = function() {
+      var r = $window.confirm('Really clear the logs?');
+      if (r === true) {
+        service.db.dropTable(service.tableName);
+        service.createTableIfNotExists();
+      }
+    };
+
+    service.export = function() {
+      service.db.selectAll(service.tableName).then(function(results) {
+        if (results.rows.length === 0) {
+          console.warn('No logs');
+          return;
+        }
+
+        var jsonData = new Blob([JSON.stringify(results.rows)], {type: 'application/json'});
+        var jsonUrl = URL.createObjectURL(jsonData);
+
+        var element = angular.element('<a/>');
+        element.attr({
+          href: jsonUrl,
+          target: '_blank',
+          download: service.tableName + '.json'
+        })[0].click();
+      });
+    };
+
+
     service.logInteraction = function(action, label, data) {
       if (!consts.logging) {
         return;
       }
       var value = data ? data.value : undefined;
       if(action.level.rank >= service.levels[consts.logLevel || 'INFO'].rank) {
         Analytics.trackEvent(action.category, action.id, label, value);
+
+        if (consts.logToWebSql) {
+          var row = {
+            userid: user,
+            time: new Date(),
+            actionCategory: action.category,
+            actionId: action.id,
+            label: _.isObject(label) ? JSON.stringify(label) : label, // zening: do we need label?
+            data: data ? JSON.stringify(data) : undefined // zening: do we log data or value?
```

Comment:
I think the `value` crap is added randomly in this commit ...  https://github.com/vega/vega-lite-ui/commit/59d59284b3c18e5ae003b9461fe37e9f3f13b386